<a href="https://colab.research.google.com/github/MarcosP7635/Energy/blob/main/JupyterNotebooks/nuclear_atlas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import subprocess
is_colab = False
if not "JupyterNotebooks" in subprocess.os.getcwd():
    is_colab = True
    !wget https://raw.githubusercontent.com/MarcosP7635/Energy/main/ImportedData/nuclear_atlas_full.txt
    parent_dir = "/content/"
else:
    parent_dir = subprocess.os.getcwd().split("JupyterNotebooks")[0]
parent_dir

'c:\\Users\\engin\\Documents\\GitHub\\Energy\\'

In [12]:
file_name = "nuclear_atlas_full.txt"
if is_colab:
    data_dir = parent_dir
else: 
    data_dir = parent_dir + "ImportedData\\"
with open(data_dir + file_name, encoding='utf-8') as file:
    data = file.read().split("\n")
[(x, data[x]) for x in range(len(data)) 
if "TABLE I" in data[x]]

[(1057, 'TABLE I')]

In [13]:
data[1057:]

['TABLE I',
 'A. K. Jain et al.',
 'Z N AX E(keV) J\x02 T1/2 Eγ(keV) λ Decay mode Key No.',
 '4 8 12Be 2251 (1) 0+ 233.0(75) ns 144(1)',
 '2251(1)',
 'E2',
 'E0',
 '%IT = 100 2013JO06',
 '2007SH34',
 '7 9 16N 120.42 (12) 0\x03 5.25(6) μs 120.42(12) %IT = 99.99960(4)',
 '%β\x03 = 0.00040(4)',
 '1983MI10',
 '1983GA18',
 '1975PA01',
 '1967BE14',
 '9 9 18F 1121.36 (15) 5+ 162(7) ns 184 E2 %IT = 100 1995TI07',
 '9 10 19F 197.143 (4) 5/2+ 89.3(10) ns 87.3',
 '197.1 E2',
 '%IT = 100 1995TI07',
 '10 9 19Ne 238.27 (11) 5/2+ 18.0(5) ns 238.3 E2 %IT = 100 1995TI07',
 '11 11 22Na 583.11 (9) 1+ 243(2) ns 583.04(10) E2 %IT = 100 1966SU07',
 '11 13 24Na 472.2074 (8) 1+ 20.18(10) ms 472.2024(8) M3 %IT = 99.95',
 '%β\x03 ≈ 0.05',
 '1961SC09',
 '1970CH37',
 '1972BR53',
 '1980JO11',
 '13 11 24Al 425.8 (1) 1+ 130.9(13) ms 425.8(1) M3 %IT = 69.6(7)',
 '%ε+%β+ = 30.4(7)',
 '%εα = 0.028(6)',
 '2007FI14',
 '2011NI18',
 '9 17 26F 643.4 (1) 4+ 2.2(1) ms %IT ≈ 82(11)',
 '%β\x03 ≈ 18(11)',
 '2013LE03',
 '11 15 26

The rows are formatted as follows:

Z N AX E(keV) J\x02 T1/2 Eγ(keV) λ Decay mode Key No.

Thus, each row begins with an atomic number and number of neutrons, so I can just find the start of a new isomer's data by looking for two consecutive integers at the beginning of a line.

If $J^\pi$ has multiple values, then this will be the end of the line, and the next line will have the half life.

In [14]:
def is_row_start(str):
   arr = str.split(" ")
   if len(arr) > 2:
      return (arr[0].isdigit() and 
              arr[1].isdigit())
   return False
is_row_start(data[1060])

True

In [15]:
row_starts = [(r, data[r])
for r in range(len(data)) 
if is_row_start(data[r])]
row_starts

[(1060, '4 8 12Be 2251 (1) 0+ 233.0(75) ns 144(1)'),
 (1066, '7 9 16N 120.42 (12) 0\x03 5.25(6) μs 120.42(12) %IT = 99.99960(4)'),
 (1072, '9 9 18F 1121.36 (15) 5+ 162(7) ns 184 E2 %IT = 100 1995TI07'),
 (1073, '9 10 19F 197.143 (4) 5/2+ 89.3(10) ns 87.3'),
 (1076, '10 9 19Ne 238.27 (11) 5/2+ 18.0(5) ns 238.3 E2 %IT = 100 1995TI07'),
 (1077, '11 11 22Na 583.11 (9) 1+ 243(2) ns 583.04(10) E2 %IT = 100 1966SU07'),
 (1078, '11 13 24Na 472.2074 (8) 1+ 20.18(10) ms 472.2024(8) M3 %IT = 99.95'),
 (1084, '13 11 24Al 425.8 (1) 1+ 130.9(13) ms 425.8(1) M3 %IT = 69.6(7)'),
 (1089, '9 17 26F 643.4 (1) 4+ 2.2(1) ms %IT ≈ 82(11)'),
 (1092, '11 15 26Na 82.5 (5) 1+ 9(2) μs 82.5 %IT = 100 1987DUZU'),
 (1094, '13 13 26Al 228.305 (13) 0+ 6.3463(67) s %ε+%β+ = 100 2011SC22'),
 (1097, '12 19 31Mg 50.1 (2) (3/2+) 16(3) ns 50.2(2) %IT = 100 1993KL02'),
 (1099,
  '12 19 31Mg 461.0 (6) (7/2\x03) 10.5(8) ns 239.9(5) %IT = 100 2005MA96'),
 (1101, '13 19 32Al 956.6 (5) (4+) 200(20) ns 221.9(3) %IT = 100 1996RO02

In [88]:
len(row_starts)

2469

For verification, the paper says they list 2469 different isomers, and that they only include isomers with a half life > 10 ns.

Source-

free PDF I put on Zotero:
https://www.zotero.org/groups/4549380/batteries/collections/VUUF9IJG/items/BWM33GUX/attachment/665IXGIJ/reader

DOI: http://dx.doi.org/10.1016/j.nds.2015.08.001)

Assuming there are no gaps in the data, the beginning of one row must be the end of the previous row. 

In [80]:
time_units = ['d', 'h', 'min', 'ns',
's', 'y', 'μs', 'ms']
def get_data(row):
   arr = row.split(" ")
   unit_index = [i for i in range(len(arr)) 
                  if arr[i] in time_units]
   if len(unit_index) == 0:
      if arr[-1] in ['to', ',']:
         #The line ends with the spin. Next line has half life
         return "Must combine with next row"
   elif len(arr) > 4 and arr[0].isdigit() and arr[1].isdigit():
         t_unit = arr[unit_index[0]]
         t_val = arr[unit_index[0]-1]
         isotope = arr[2]
         E_total = arr[3]
         return (isotope, E_total, 
              t_val, t_unit)
   return "Improperly Formatted"

data_tuples = [get_data(" ".join(data[row_starts[index][0]:row_starts[index+1][0]]))
                        for index in range(len(row_starts)-1)]
data_tuples

[('12Be', '2251', '233.0(75)', 'ns'),
 ('16N', '120.42', '5.25(6)', 'μs'),
 ('18F', '1121.36', '162(7)', 'ns'),
 ('19F', '197.143', '89.3(10)', 'ns'),
 ('19Ne', '238.27', '18.0(5)', 'ns'),
 ('22Na', '583.11', '243(2)', 'ns'),
 ('24Na', '472.2074', '20.18(10)', 'ms'),
 ('24Al', '425.8', '130.9(13)', 'ms'),
 ('26F', '643.4', '2.2(1)', 'ms'),
 ('26Na', '82.5', '9(2)', 'μs'),
 ('26Al', '228.305', '6.3463(67)', 's'),
 ('31Mg', '50.1', '16(3)', 'ns'),
 ('31Mg', '461.0', '10.5(8)', 'ns'),
 ('32Al', '956.6', '200(20)', 'ns'),
 ('32Si', '5581', '27(2)', 'ns'),
 ('33Si', '1434.9', '10.2(3)', 'ns'),
 ('34Al', '0+X', '26(1)', 'ms'),
 ('34Cl', '146.36', '31.99(3)', 'min'),
 ('37K', '1380.25', '10.4(5)', 'ns'),
 ('38Cl', '671.365', '715(3)', 'ms'),
 ('38K', '130.15', '924.46(14)', 'ms'),
 ('38K', '3458.14', '21.95(11)', 'μs'),
 ('40K', '1643.639', '0.336(12)', 'μs'),
 ('42Sc', '616.28', '61.7(4)', 's'),
 ('43S', '320.9', '0.20(+14\x037)', 'μs'),
 ('43K', '738.1', '200(5)', 'ns'),
 ('43Sc', '151.4', 

In [81]:
set([x[3] for x in data_tuples])

{'d', 'h', 'min', 'ms', 'ns', 'r', 's', 'y', 'μs'}

In [87]:
bad_rows = [r for r in range(len(data_tuples)) if type(data_tuples[r]) == str]
bad_rows

[450, 469, 673]

In [86]:
data_tuples[bad_rows[0]]

'Improperly Formatted'

In [53]:
data[450:1245]

['21 33 54Sc 110.0 (10) (4,',
 '5) +',
 '2.77(2) μs 110 E2 %IT = 100 2010CR02',
 '23 31 54V 108.0 (10) (5)+ 0.9(5) μs 108 E2 %IT = 100 1998GR14',
 '26 28 54Fe 6527.1 (11) 10+ 364(7) ns 146.2 (2)',
 '3577.6']

In [46]:
for x in bad_rows:
    print(data[x])

21 33 54Sc 110.0 (10) (4,
21 35 56Sc 0+X (5+,
35 38 73Br 240.50 (13) (3/2
29 46 75Cu 61.7 (4) (1/2
29 46 75Cu 66.2 (4) (1/2
41 44 85Nb 69+Y (1/2,
41 45 86Nb 49.96+X (1+,
35 53 88Br 270.1 (5) (34
43 45 88Tc X (5+,
43 45 88Tc Y (2+,
35 60 95Br 537.9 (5) (3/2
46 50 96Pd 8385.1 (8) (1516
37 60 97Rb 77+X (1/2,
39 59 98Y 374.91 (12) (12
41 59 100Nb 34.3+X (45
39 62 101Y 1205.5 0.860(+9080)
50 52 102Sn 1969+Y (6+) 0.8(3) μs* 1996LI50
44 61 105Ru 163.815 (15) 3/2+,
51 63 114Sb 45.86 (2) (2+),
46 71 117Pd 203.3 (3) (9/2,
47 71 118Ag 279.37 (20) (2+,
48 71 119Cd 228.27 (9) (7/2
48 73 121Cd 314.50 (6) (5/2+,
49 72 121In 2448.0 (10) (21/2−) 17(2)μs # 214? 2010RE01
48 75 123Cd 263.867 (20) (5/2+,
57 73 130La 110.44 (9) (1+, 2,
53 78 131I 4307.4 (6) (31/2
61 74 135Pm# 0+X (3/2+,
59 90 149Pr 58.5 (3) (7/2,
71 79 150Lu 22 (6) (1
59 92 151Pr 35.10 (10) (1/2+,
70 93 163Yb 58.1 (3/2
71 95 166Lu 83.50 (10) (5,6,
75 94 169Re 0+X (1/2+,
75 104 179Re 5408.0 (5) (47/2,
73 107 180